src/block-management/block-stream.js

```diff
@@ -0,0 +1,169 @@
+"use strict";
+
+var BlockNotifier = require("./block-notifier.js");
+var Notifier = require("./notifier.js");
+var validateBlock = require("./validate-block.js");
+
+/**
+ * This object will emit a stream of new blocks, after handling backfills and reorgs.
+ * 
+ * @typedef transport
+ * @type object
+ * @property {function(string, boolean, function(object):void):void} getBlockByHash
+ * 
+ * @param {Transport} transport
+ * @param {Notifier} blockNotifier - A Notifier that will fire on new block candidates.
+ */
+function BlockStream(transport, pollingIntervalMilliseconds) {
+  Notifier.call(this);
+
+  var blocks = {};
+  var headBlock = null;
+  var sidelinedBlocks = {};
+  var blockRetentionLimit = 100;
+  var blockNotifier = new BlockNotifier(transport, pollingIntervalMilliseconds);
+
+  this.destroy = function () {
+    this.unsubscribeAll();
+    blockNotifier.destroy();
+  }.bind(this);
+
+  var shouldKeepBlock = function(block) {
+    if (!block) return true;
+    var blockNumber = parseInt(block.number);
+    if (isNaN(blockNumber)) return true;
```

Comment:
Good catch, those early returns should both be `false`.  I think I originally had the function `shouldNotKeepBlock` and when I inverted it I missed those two early return statements.